src/glang/lib/std/rb_tree.c3

```diff
@@ -0,0 +1,481 @@
+        // red-and-black.
+        w.colour = RED();
+        return make_optional(&y);
+    }
+
+    // Case 3: w is black with red close and black distant child.
+    if w:get_child(x_side):is_red() and w:get_child(w_side):is_black() {
+        // Make w's red child w's new parent, and colour it black, such that x's
+        // sibling is now black. w must in turn become red, to compensate.
+        self:rotate(&w, w_side);
+        w.colour = RED();
+        w.parent:colour_black();
+
+        // x remains in place, and we get case 4.
+        return x;
+    }
+
+    // Case 4: w is black and w's distant child is red.
+    runtime_assert(w:get_child(w_side):is_red());
+
+    // Essentially, we make w y's parent, which means that there is an
+    // additional black node now in the path of x. This allows us to remove the
+    // double black. The root of the sibling tree is now w's red child, which
+    // we colour black to compensate.
+    self:rotate(&y, x_side);
+
+    // w takes y's colour, so the colour of this subtree's root is preserved.
+    w.colour = y.colour;
+
+    // We have removed one black node from w's side of the tree (w itself), so
+    // we need to make its red child black to compensate (as it is now the root
+    // of that subtree). On the other side, the new root should also be black,
+    // as it absorbs the "double black" from x.
+    w.left:colour_black();
+    w.right:colour_black();
+
+    // Remove the double black; we are done.
+    return self.root;
+}
+
+function [K] fix_after_delete : (
+    self : RB_Tree<K>&, x : Optional<RB_Node<K>&>, y : Optional<RB_Node<K>&>
+) -> void = {
+    // x is "double black" inside the loop.
+    while x != self.root and x:is_black() {
+        // x might be a leaf node (nil node), so take its parent, y as an
+        // argument. y is an optional, as the root node has no parent. However,
+        // it should have a value in all other cases.
+        x = self:fix_after_delete_impl(x, &y:data());
+
+        // In cases 1 and 3, x is unchanged, so it could still be an empty
+        // optional, representing a leaf node (nil node). In the other cases, x
+        // moves up the tree, so we must update y.
+        if x:has_value() {
+            y = x:data().parent;
+        }
+    }
+
+    // If x was the root, then it must always be black. If x was
+    // "red-and-black" (i.e. x was originally red), we can colour it black to
+    // resolve the issue.
+    x:colour_black();
+}
+
+function [K] delete_impl : (self : RB_Tree<K>&, u : RB_Node<K>&) -> void = {
+    // x is the "double black" or "red-and-black" node which we use to fix any
+    // red-black tree violations that arise from the deletion. y is its parent,
+    // as x might be a leaf node (nil node), which is represented as an empty
+    // optional.
+    let x : Optional<RB_Node<K>&> = make_optional<RB_Node<K>&>();
+    let y : Optional<RB_Node<K>&>;
```

Comment:
I've made it an empty optional, although that doesn't really do much for us, because what we really want is to check that it has been assigned to. Did you have anything else in mind?